File: src/modules/Pieces/Utils/CastlePiece/CastlePiece.cpp
#pragma once

#include "CastlePiece.h"

#include "../../Utils/RawPosition/RawPosition.h"
#include "../../../Common/InstanceOf/InstanceOf.h"
#include "../../../Common/PieceColor/PieceColor.h"

CastlePiece::CastlePiece(PieceParams* params):
    HasMoved(params)
{}

bool CastlePiece::isTreatened() {
    vector<RawPosition> enemyMoves = this->getEnemyMoves(true);

    bool isTreatened = false;

    for (int i = 0; i < enemyMoves.size(); i++) {
        RawPosition* rawPosition = &enemyMoves.at(i);
        RawPosition* thisPosition = this->pieceImpl->getRawPosition();

        if (
            rawPosition->getColumn() == thisPosition->getColumn()
            && rawPosition->getRow() == thisPosition->getRow()
        ) {
            isTreatened = true;

            break;
        }
    }

    return isTreatened;
}

int CastlePiece::getCorrespondingRow() {
    return this->pieceImpl->getColor() == PieceColor::WHITE
        ? 7
        : 0;
}

/**
 * Determines wheter a row composed of positions
 * is empty and also is not treatened by enemy
 * pieces.
 */
bool CastlePiece::checkIfEmptyAndSafe(int fromInclusive, int toInclusive) {
    GameController* gameController = this->pieceImpl->getGameController();
    unique_ptr<GameState>& gameState = gameController->gameControllerImpl->getGameState();

    int min = std::min(fromInclusive, toInclusive);
    int max = std::max(fromInclusive, toInclusive);

    for (int i = min; i <= max; i++) {
        RawPosition position = RawPosition(
            i,
            this->getCorrespondingRow()
        );

        GetValidPieceResponse pieceResponse = gameState->gameStateImpl->getValidPiece(
            position.getColumn(),
            position.getRow()
        );
        
        if (InstanceOf::check<Piece<RawPosition>>(pieceResponse.piece)) return false; // Because not empty

        vector<RawPosition> enemyMoves = this->getEnemyMoves(true);

        for (int j = 0; j < enemyMoves.size(); j++) {
            RawPosition* possibleMove = &enemyMoves.at(j);

            if (
                possibleMove->getColumn() == position.getColumn()
                && possibleMove->getRow() == position.getRow()
            ) {
                return false; // Because not safe
            }
        }
    }

    return true;
}

vector<RawPosition> CastlePiece::getEnemyMoves(bool disableCastle = false) {
    vector<RawPosition> enemyMoves = vector<RawPosition>();

    GameController* gameController = this->pieceImpl->getGameController();
    vector<Piece<RawPosition>*>* piecePtrs = gameController->gameControllerImpl->getPiecePtrs();

    for (Piece<RawPosition>* piece: *piecePtrs) {
        // Search for enemy pieces, discard friendly ones
        if (piece->pieceImpl->getColor() == this->pieceImpl->getColor()) continue;

        if (InstanceOf::check<Rook>(piece)) {
            vector<RawPosition> possibleMoves = static_cast<Rook*>(piece)->getPossibleMoves(disableCastle);
            
            for (int i = 0; i < possibleMoves.size(); i++) {
                RawPosition possibleMove = possibleMoves.at(i);

                enemyMoves.push_back(possibleMove);
            }

            continue;
        }


        if (InstanceOf::check<King>(piece)) {
            vector<RawPosition> possibleMoves = static_cast<King*>(piece)->getPossibleMoves(disableCastle);
            
            for (int i = 0; i < possibleMoves.size(); i++) {
                RawPosition possibleMove = possibleMoves.at(i);

                enemyMoves.push_back(possibleMove);
            }

            continue;
        }

        vector<RawPosition> possibleMoves = piece->getPossibleMoves();

        for (int i = 0; i < possibleMoves.size(); i++) {
            RawPosition possibleMove = possibleMoves.at(i);

            enemyMoves.push_back(possibleMove);
        }
    }

    return enemyMoves;
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Utils/CastlePiece.ts
import { King, Rook } from "../Kinds";
import { Piece } from "../Piece";
import { Position } from "../types";
import { HasMoved } from "./HasMoved";

export abstract class CastlePiece extends HasMoved {
    protected get correspondingRow(): 7 | 0 {
        return this.color === "white" ? 7 : 0;
    }

    /**
     * Determines wheter a row composed of positions
     * is empty and also is not treatened by enemy
     * pieces.
     */
    protected checkIfEmptyAndSafe(fromInclusive: number, toInclusive: number): boolean {
        const { board } = this.controller.gameState;

        const min = Math.min(fromInclusive, toInclusive);
        const max = Math.max(fromInclusive, toInclusive);

        for (let i = min; i <= max; i++) {
            const position: Position = {
                row    : this.correspondingRow,
                column : i,
            }

            const piece: Piece | null = board
                ?.[position.row]
                ?.[position.column];
            
            if (piece instanceof Piece) return false; // Because not empty

            const notSafe = this.getEnemyMoves({ disableCastle : true })
                .findIndex(possibleMove =>
                    possibleMove.row === position.row
                    && possibleMove.column === position.column
                ) !== -1;

            if (notSafe) return false;
        }

        return true;
    }

    protected getEnemyMoves({ disableCastle } = { disableCastle : false }): Position[] {
        const enemyMoves: Position[] = [];
        
        for (const piece of this.controller.gameState.board.flat()) {
            if (!(piece instanceof Piece)) continue;
            
            // Search for enemy pieces, discard friendly ones
            if (piece.color === this.color) continue;

            if (piece instanceof Rook || piece instanceof King) {
                const possibleMoves = piece.getPossibleMoves({ disableCastle });
                
                enemyMoves.push(...possibleMoves);
                continue;
            }

            const possibleMoves = piece.getPossibleMoves();

            enemyMoves.push(...possibleMoves);
        }

        return enemyMoves;
    }

    public isTreatened(): boolean {
        const enemyMoves: Position[] = this.getEnemyMoves({ disableCastle : true });

        const isTreatened = enemyMoves.findIndex(possibleMove =>
            possibleMove.row === this.position.row
            && possibleMove.column === this.position.column
        ) !== -1;

        return isTreatened;
    }
}
